### PR TITLE
Make GRIB API optional.

### PR DESCRIPTION
**Pull request by rhattersley**
NB. This adds a new entry to the test matrix which runs the standard test suite against an installation with a reduced set of dependencies. Currently this is only missing ecmwf_grib but I would expect other dependencies could be removed once the concept is proven.

### CODE REVIEW

**Comment by rhattersley** [comment]
NB. The examples and doctests are only tested with the **full** set of dependencies. The examples will certainly fail if ecmwf_grib is missing - I'm not sure about the doctests.

**Comment by rhattersley** [comment]
> The examples will certainly fail if ecmwf_grib is missing - I'm not sure about the doctests.

Whether this is the behaviour we want is debatable! :wink:

File: lib/iris/tests/runner/_runner.py [hunk 1]
```diff
@@ -181,6 +181,10 @@ def run(self):
         args = ['', None, '--processes=%s' % n_processors,
                 '--verbosity=2', regexp_pat,
                 '--process-timeout=250']
+        try:
+            import gribapi
+        except ImportError:
+            args.append('--exclude=^grib$')
```
Comment (rhattersley) [inline lib/iris/tests/runner/_runner.py]: NB. This matches the directory name "grib" when nose is doing test discovery and so excludes all of the tests under `lib/iris/tests/unit/fileformats/grib/...`. The alternative was to modify all those tests to add `skip_grib`.... :unamused:

File: lib/iris/fileformats/__init__.py [hunk 2]
```diff
@@ -74,10 +77,11 @@ def _pp_little_endian(filename, *args, **kwargs):
...
 # priority to avoid collateral damage from false positives.
-FORMAT_AGENT.add_spec(
-    FormatSpecification('GRIB', MagicNumber(100),
-                        lambda header_bytes: b'GRIB' in header_bytes,
-                        grib.load_cubes, priority=1))
+if grib is not None:
+    FORMAT_AGENT.add_spec(
+        FormatSpecification('GRIB', MagicNumber(100),
+                            lambda header_bytes: b'GRIB' in header_bytes,
+                            grib.load_cubes, priority=1))
```
Comment (pelson) [inline lib/iris/fileformats/__init__.py]: Nice if this still knew what a grib file was, and told us we didn't have grib installed.

File: lib/iris/fileformats/grib/__init__.py [hunk 3]
```diff
@@ -31,13 +31,11 @@
 
 import biggus
 import cartopy.crs as ccrs
+import gribapi
 import numpy as np
 import numpy.ma as ma
 import scipy.interpolate
 
-import iris.proxy
-iris.proxy.apply_proxy('gribapi', globals())
```
Comment (pelson) [inline lib/iris/fileformats/grib/__init__.py]: Can we deprecate iris.proxy for 2.0

File: lib/iris/io/__init__.py [hunk 4]
```diff
@@ -234,8 +234,13 @@ def _check_init_savers():
...
                         "dot": iris.fileformats.dot.save,
-                        "dotpng": iris.fileformats.dot.save_png,
-                        "grib2": iris.fileformats.grib.save_grib2})
+                        "dotpng": iris.fileformats.dot.save_png})
+        try:
+            import gribapi
+        except ImportError:
+            pass
+        else:
+            _savers['grib2'] = iris.fileformats.grib.save_grib2
```
Comment (pelson) [inline lib/iris/io/__init__.py]: Ditto.

File: .travis.yml [hunk 5]
```diff
@@ -52,7 +53,11 @@ install:
   - conda config --add channels scitools
   # Explicitly add missing libgfortran dependency for numpy.
   # See SciTools/iris#1777 and ContinuumIO/anaconda-issues#445.
-  - conda install --quiet --file conda-requirements.txt libgfortran
+  - if [[ "$TEST_MINIMAL" == true ]]; then
+      conda install --quiet --file minimal-conda-requirements.txt libgfortran;
```
Comment (pelson) [inline .travis.yml]: We can do:

```
conda install --quiet --file conda-requirements.txt --file conda-optional-requriements.txt
```

Comment (pelson) [inline .travis.yml]: The libgfortran isn't necessary any more.

File: lib/iris/fileformats/__init__.py (hunk 2 above)
Comment (rhattersley) [inline lib/iris/fileformats/__init__.py]: :+1: nice idea

File: .travis.yml [hunk 6]
```diff
@@ -13,6 +13,7 @@ sudo: false
 
 env:
   - TEST_TARGET=default
+  - TEST_TARGET=default TEST_MINIMAL=true
   - TEST_TARGET=coding
```
Comment (pelson) [inline .travis.yml]: Maybe we can make as many of these "minimal" too - it might speed up some of our builds a little.

File: .travis.yml (hunk 5 above)
Comment (rhattersley) [inline .travis.yml]: :+1: for the two `--file` approach

File: minimal-conda-requirements.txt [hunk 7]
```diff
@@ -0,0 +1,27 @@
+# Use this file to create a conda environment using:
```
Comment (pelson) [inline minimal-conda-requirements.txt]: Suggest changing the directory structure to:

```
requirements/
             core.txt
             development.txt
             optional.txt
```

File: lib/iris/fileformats/grib/__init__.py (hunk 3 above)
Comment (rhattersley) [inline lib/iris/fileformats/grib/__init__.py]: Yes ... but I suggest we do that in a separate PR.

File: minimal-conda-requirements.txt (hunk 7 above)
Comment (rhattersley) [inline minimal-conda-requirements.txt]: conda doesn't (currently) let you use multiple `--file` arguments so I've stuck with the existing two-file structure.

Comment (pelson) [inline minimal-conda-requirements.txt]: https://github.com/conda/conda/pull/1633 refers.

File: .travis.yml (hunk 6 above)
Comment (rhattersley) [inline .travis.yml]: Good idea. I'd suggest it as a follow-up activity as it doesn't look like it makes a huge difference right now, e.g. 2 mins out of 15-21 mins.

**Comment by rhattersley** [comment]
:tada: 

Thanks @pelson :grin:

File: .travis.yml (hunk 6 above)
Comment (QuLogic) [inline .travis.yml]: The majority of the time is the conda dep solver after all.
